### ISSUE TITLE
Upgrade Management Page

### ISSUE THREAD

**Issue by snkaupe** [joshmue/scs-status-page-web#42]
In our rush to finish the CUD features/the management page for R7, some things were implemented a bit haphazardly due to time constraints. Now, we have the time to clean these up or replace them with an improved version.

Together with @joshuai96 , we decided we could upgrade the interaction concept for the page, while also allowing it to perform a greater breadth of editing tasks. To do so, we will replace the management view with detail pages for the entities managed by the status API server, with editing functionality build into said pages. This will allow the web frontend to edit not just incidents, incident updates and maintenance events, but also components, severity names, etc.

**Comment by snkaupe** [comment]
What has happened so far?

My initial plan was to create a new, reusable component to implement the edit functionality needed on the detail view pages. In the initial vision, this component would have received a reference to the object to be edited as well as a list of the properties one wants to edit, i.e. `beganAt` and `displayName` on an `Incident`. The component would then automatically select a fitting `<input>` element for this field and allow the user to switch between display and editing mode. The component would also include the ability to detect changes to the selected properties, to allow the display of reminders about unsaved changes. Ideally, this component could then have been extracted into a standalone Angular component and released separately under a FOSS license, giving back to the FOSS community.

Idealized snippet to illustrate the idea:

```html
<SwitchableEdit
    [source]="someIncident"
    [properties]="displayName,beganAt"
></SwitchableEdit>
```

This would create a part of the page displaying the incident's name and start date/time, as well as a button to switch it into editing mode. Clicking this button would exchange the static display elements with `<input>` fields of proper types, derived from the datatypes of the listed properties (i.e. a `<input type="text">` for the name, a `<input type="datetime-local">` for the date, etc.). Of course, additional properties would be needed to apply styling information.

This way, the user could have enabled editing on whatever properties of an entity they want to change, perform the changes and then save them using a single save button on the page. This would cause all `SwitchableEdit`s in an *active* state (i.e. potentially with unchanged changes) to run validation functions that detect invalid input and then, if all checks clear, the surrounding component could perform the call to the API server and save the changed object.

I implemented an early prototype, retrieving properties, determining data types and mapping them onto fitting input elements. However, problems arose somewhat quickly once constraints not contained in the mere type information had to be considered. As an example, an incident's `beganAt` date must never occur before its `endedAt` date. Or the `displayName` property must not be empty. These cannot be retrieved from type information alone and must be provided on a different way.

Inspired by the use of annotations in Java, I first considered decorators added to the fields of the model classes, something like this:
```typescript
export interface Incident { 
    @NotEmpty
    displayName?: string;
    @Multiline
    description?: string;
    readonly updates?: Array<number>;
    affects?: Array<Impact>;
    @NotEmpty
    beganAt?: string | null;
    @After("beganAt")
    endedAt?: string | null;
    phase?: PhaseReference;
}
```
However, I had to discover that decorators cannot be added to Typescript interfaces. Furthermore, the model classes are currently generated by an OpenAPI generator and adding support for these decorators (or switching to classes instead of interfaces) would have required us to either give up the automatic generation, modify the generator (requiring either a custom fork or a whole other, different plan on how to express these constraints in an OpenAPI specification) or add the decorators in a custom post-generation step, which would require us to parse the generated Typescript sources (as well as yet another way to express the constraints).

Therefore, expressing these constraints could only have happened in the Angular sources themselves. I experimented with different ways to do so, e.g. in the HTML itself:

```html
<SwitchableEdit
    [source]="someIncident"
    [properties]="beganAt|NotEmpty,endedAt|After(beganAt)"
></SwitchableEdit>
```

I also considered a specification provided in a field of the detail view's Typescript source and an external constraint file, loaded at the beginning of the application's run and mapping constraints onto properties, like this:

```yaml
entities:
  Incident:
    displayName:
      - NotEmpty
    description:
      - Multiline
    beganAt:
      - NotEmpty
    endedAt:
      - After: beganAt
  Component:
  [...]
```

However, the component became increasingly unwieldy and less ergonomic to use and the code to parse what could basically be considered a constraint DSL was fickle. Additionally, the references to other entities turned out to be a bigger problem than initially thought, as the `SwitchableEdit` needed a way to access a list of entities of the correct type, information on how to display them and information on the constraints that apply to these lists as well (i.e. "no component may occur more than once in the `affects` field of an `Incident`).

Alternatively, I could have scrapped the idea of automatic checking of dirty status and input validity, but these features would then have to be implemented in every component that is making use of the `SwitchableEdit` by way of manually written validation methods. These, in turn, would either have to be handed to the edit in the HTML source of would have needed access to the input fields in the `SwitchableEdit` itself, making the whole approach meaningless.

Overall, I had to take a step back, look at the work done so far and make the decision that this approach would not work due to the aforementioned problems.

**Comment by snkaupe** [comment]
The new plan is much simpler: We (probably) still have a component called `SwitchableEdit`, but it only provides the boilerplate for switching the elements. Said elements themselves are just added as the edit's children in the HTML and fitted with the appropriate CSS classes to enable switching. Validation is provided by the component for the surrounding detail view.

**Comment by garloff** [comment]
Alright, keep us posted how it goes, please!

**Comment by snkaupe** [comment]
There seems to be a slight problem with the way data is being loaded at the moment. When reloading the details page for an incident, any updates the incident might have are missing. I suspect this is due to us currently not tracking the loading progress for updates, instead declaring loading to be done once everything else is done loading. This was fine so far, as we'd always start on the front page, but it might be a slight problem once we start editing right on an incident's detail page.

Proposed solution: Keep track for which incidents we've already retrieved all updates from the API server and only mark loading as done once we've checked updates for every incident. This necessitates some changes to the `DataService`. I'll probably mark this as a lower priority issue for the moment.

**Comment by snkaupe** [comment]
Contrary to my earlier post, I decided to perform a quick hotfix for the issue. It might still contain some potential for a false positive when it comes to loading completion, but I'd have to invest some time to think it through thoroughly and might need to perform some tests. For now, this is sufficient.

**Comment by snkaupe** [comment]
Status update:

I'm mostly done with the detail/edit view for the incidents. It took a bit longer than expected because it is the most complicated of the detail pages, but I've spend some time separating reusable elements into their own components (error display, the control buttons on top, service to centralize control of icons) and generalized utilities (data conversion, error checking and handling) so that creating the other entity pages will be much quicker. I'm currently working on the dialogues to add updates and affected components.

I've also went and restructured the project. Formerly, all components, pages, pipes and services simply shared a flat directory, but this was becoming quick unhandy to deal with. So, into their own subfolders they went.

Follow PR #43 to see my progress.

**Comment by snkaupe** [comment]
Due to time constraints, it won't be possible to create editable detail pages for all entities handled by the API server and we will have to scale back. The new target is now feature parity with the old management page, which means completion of the detail view for incidents.

Luckily, this is the most complicated of all detail pages and should prove a handy template for any later work happening on this project that aims to implement the original goals for this ticket.

To do this, one would need to go and add the following:

- Some kind of overview, listing all available entities so that they can be clicked on for editing (subpage or a menu with openable submenus)
- An editable detail page for each entity meant to be editable via the frontend (using the `incident-detail-view` as a template)

The detail view for incidents contains a number of components that can be reused, just like the general setup of the page.

**Comment by snkaupe** [comment]
Everything is implemented and we now have feature parity with the old management page. Huzzah!

PR just needs someone to review/approve it and then the status page hosted on our SCS infrastructure can be updated.
